packages/@react-native-windows/cli/src/generator-windows/index.ts

```diff
@@ -118,11 +118,33 @@ export async function copyProjectTemplateAndReplace(
   if (options.experimentalNuGetDependency) {
     console.log('Using experimental NuGet dependency.');
   }
+
+  let realProjectType = projectType;
+
   if (options.useWinUI3) {
     console.log('Using experimental WinUI3 dependency.');
+    if (projectType === 'lib') {
+      throw new CodedError(
+        'IncompatibleOptions',
+        'WinUI 3 project template only supports apps at the moment',
+        {
+          detail: 'useWinUI3 and lib',
+        },
+      );
+    } else if (language !== 'cs') {
+      throw new CodedError(
+        'IncompatibleOptions',
+        'WinUI 3 project template only support C# at the moment',
+        {
+          detail: 'useWinUI3 and cpp',
+        },
+      );
+    }
+
+    realProjectType += '-reunion';
```

Comment:
This sidesteps the mechanisms other flavors (e.g. binary distribution) use to specialize the generated template, and makes WindowsAppSDK unique to maintain and test. Can we do something cleaner here?